### ISSUE TITLE
ib_insync in docker

### ISSUE THREAD

**Issue by roa07620** [erdewit/ib_insync#261]
I am trying to use ib_insync with the docker image "mvberg/ib-gateway-docker".
IBController runs successfully, Gateway is started etc.
However, ib_insync does not work. The call ib.connect('ip of docker',4001) times ou, even with timeout=None. Any ideas how this issue can be solved? Thanks!

**Comment by rbrok** [comment]
Could you please provide more information? Can you connect via vncviever to port 5901 or 5900?
Please check the network config first.

**Comment by goodboy** [comment]
@roa07620 what ports does the container proxy because 4001 isn't the default afaik for `ib-gw`.

**Comment by goodboy** [comment]
Also linking to the `Dockerfile` would help :eyes: 

https://github.com/mvberg/ib-gateway-docker

**Comment by goodboy** [comment]
Looks like [4001 is supposedly forced in the `IBController` config](https://github.com/mvberg/ib-gateway-docker/blob/master/ib/IBController.ini#L26) and `IBController` seems [to set it's own port](https://github.com/mvberg/ib-gateway-docker/blob/master/ib/IBController.ini#L59) for something.

Also if you just [read the docs](https://github.com/mvberg/ib-gateway-docker#expected-output) it says:
>You will now have the IB Gateway app running on port 4003 and VNC on 5901.

So, yeah.

**Comment by goodboy** [comment]
Also for anyone interested in the technical deats (I know I was), compose just [proxies the port](https://github.com/mvberg/ib-gateway-docker/blob/master/docker-compose.yml#L6) wholesale for this project mentioned by the OP. Which I guess means that must be some default in `IBController`?

Interestingly, `IBController` docs make no mention of this port, nor does it seem to be searchable in the codebase.
However, `4001` [is mentioned as the default in the config](https://github.com/ib-controller/ib-controller/blob/99410bd0782e1634bf7f71f198e175841e7ebb21/resources/IBController.ini#L246). 

Digging further you find out there's been a nearly half-year long issue that was made on this :rofl: 
https://github.com/ib-controller/ib-controller/issues/109

Which leads to **these 3** different additional docker repos (not including the one mentioned by OP) aimed to solve this issue:
- https://github.com/ibizaman/docker-ibcontroller
- https://github.com/buzzcloudau/docker-ibcontroller
- https://github.com/ryankennedyio/ib-docker

Finally the author of the last one [says this](https://github.com/ib-controller/ib-controller/issues/109#issuecomment-253949284):
> I believe the issues everyone is having is that within docker, the IB process is listening on :::4001. 4001 is then mounted from host into docker, but no connections can be made. This is because the process within docker must be available at 0.0.0.0:4001.
I'll probably do a bit of a tidy up and a blog post about it in a couple of days, once I remove the dependancy on hardcoding 4003.

That blog post unfortunately looks to no longer be hosted.
Another person [suggests this all be summarized in another issue](https://github.com/ib-controller/ib-controller/issues/109) which I guess never happened :crying_cat_face: (until now :surfing_man:).

Finally it looks as though the last repo / project attempt [worked for someone](https://github.com/ib-controller/ib-controller/issues/109#issuecomment-284843534).

Digging into that repo you can find some explanation of the mystery that is setting up a docker port forward from a container:
https://github.com/ryankennedyio/ib-docker#ports

So it seems to be some `IBController` problem to do with getting the port proxied which is solved with a [`socat` command](https://github.com/ryankennedyio/ib-docker/blob/master/runscript.sh#L11).

So, it looks to me as though `IBController` seems some what of a :exploding_head: project and I can see why OP may have run into problems with this. 

In summary, it looks like the project/repo that that OP mentioned also [does this `socat` thinger](https://github.com/mvberg/ib-gateway-docker/blob/master/runscript.sh#L11) so port `4003` should in theory work.

Hopefully this summary prevents any confusion in the future since I think `ib_insync` is a good place for people to come for answers regarding the dark magic that is IB's api :smiley:

**Comment by erdewit** [comment]
The answer seems to be to use port 4003. Note that IBController is an abandoned project, it lives on as IBC:

https://github.com/IbcAlpha/IBC
